graphReconstruct/graphReconstruct.go

```diff
@@ -2,230 +2,91 @@ package graphReconstruct
+
+//seqOfPath takes in a graph and a path specified by the Node IDs and returns the seq of the path through the graph
+func seqOfPath(g *simpleGraph.SimpleGraph, path []uint32) []dna.Base {
+	var seq []dna.Base
+	for n := 0; n < len(g.Nodes); n++ {
+		for p := 0; p < len(path); p++ {
+			if g.Nodes[n].Id == path[p] {
+				seq = append(seq, g.Nodes[n].Seq...)
 			}
 		}
-		scrapNew[i] = sum
-	}
-	if node.Up != nil {
-		BubbleUp(node.Up, node, scrapNew)
-	} else if node.Up == nil {
-		tot = scrapNew[0] + scrapNew[1] + scrapNew[2] + scrapNew[3]
-		node.Scrap = tot
 	}
+	return seq
 }
 
-//fix each node and return the probabilities for each base at that site
-func FixFc(root *expandedTree.ETree, node *expandedTree.ETree) []float64 {
-	ans := []float64{0, 0, 0, 0}
+//PathFinder takes a graph and returns the most likely path through that graph after checking all possible paths from the first node to the last
+func PathFinder(g *simpleGraph.SimpleGraph) ([]uint32, float32) {
+	var finalPath []uint32
+	var finalProb float32
+	var tempPath = make([]uint32, 0)
 
-	for i := 0; i < 4; i++ {
-		scrap := []float64{0, 0, 0, 0} //checking one base at a time each time you call BubbleUp
-		scrap[i] = node.Stored[i]
-		if node.Up != nil {
-			//Bubble up the tree using the memory of the previous node in relation to changing position taking in probabilities of bases
-			//(node will be BubbleUp prevNode and node.Up will be the node being operated on)
-			BubbleUp(node.Up, node, scrap) //node becomes PrevNode and scrap is set to one value of prevNode.Stored in BubbleUp
-			ans[i] = root.Scrap            //root.Stored has previously assigned values (SetInternalState), you want to use whatever is returned by BubbleUp instead
-		} else if node.Up == nil {
-			ans[i] = root.Stored[i]
+	for n := 0; n < len(g.Nodes); n++ {
+		if g.Nodes[n].Id == 0 {
+			finalProb, finalPath = bestPath(g.Nodes[n], 1, tempPath)
 		}
 	}
-
-	return ans
+	return finalPath, finalProb
 }
 
-//called by reconstructSeq.go on each base of the modern (leaf) seq. Loop over the nodes of the tree to return most probable base to the Fasta
-func LoopNodes(root *expandedTree.ETree, position int) {
-	internalNodes := expandedTree.GetBranch(root) //does this logic look like it will get leaves as well? We may want to use GetLeaves, which looks like it gets everything
-	SetState(root, position)
-	for k := 0; k < len(internalNodes); k++ {
-		fix := FixFc(root, internalNodes[k])
-		yHat := Yhat(fix)
-		internalNodes[k].Fasta.Seq = append(internalNodes[k].Fasta.Seq, []dna.Base{dna.Base(yHat)}...)
+//bestPath is the helper function for PathFinder, and recursively traverses the graph depth first to determine the most likely path from start to finish
+func bestPath(node *simpleGraph.Node, prevProb float32, prevPath []uint32) (prob float32, path []uint32) {
+	var tempProb float32 = 0
+	var tempPath []uint32
+	var finalProb float32
+	var finalPath []uint32
+
+	tempPath = append(tempPath, prevPath...)
```

Comment:
may be a better way to do this than appending. I recommend either using the builtin copy function, or declaring the capacity of tempPath as len(prevPath)

Reply:
On second thought, do you even need tempPath? Can you just append the next node to the prevPath (or just rename it "path") and input that to bestPath? Since you only append nodes to the path and don't change earlier nodes in the path, the underlying path should be memory stable... i think. Recursion hurts my brain so the only way to be sure would be to test it, but I think this can be done with just 1 underlying slice in memory.

Reply:
I made a quick program to test this

https://play.golang.org/p/YJftVT3KJCH

I believe this will work, but in order to save yourself from constantly shifting memory during append functions you might consider setting the capacity of 'path' to len(graph.Nodes). It will be more than you need, but should make things alot faster.

Reply:
so i totally agree about the rest of it and i'm going to push an update, but i did a test to confirm and setting a length longer than necessary just adds a repeat of the sequence for the node labelled with 0 for its id.

Reply:
Don't make length equal to len(graph.Nodes), make capacity equal to len(graph.Nodes). This would look like `path := make([]uint32, 0, len(graph.Nodes)). This way you start with an empty slice, but can add to it without needing to allocate more memory with append.